tests/std/tests/GH_000625_vector_bool_optimization/test.cpp

```diff
@@ -253,10 +257,748 @@ void test_huge_vector_bool() {
+        vector<bool> dest(8, false);
+        const auto res_copy = copy(next(source.begin(), 7), next(source.begin(), 10), next(dest.begin(), 2));
+        assert(dest == result);
+        assert(res_copy == next(dest.begin(), 5));
+
+        vector<bool> dest_n(8, false);
+        const auto res_copy_n = copy_n(next(source.begin(), 7), 3, next(dest_n.begin(), 2));
+        assert(dest_n == result);
+        assert(res_copy_n == next(dest_n.begin(), 5));
+
+        vector<bool> dest_backward(8, false);
+        const auto res_copy_backward =
+            copy_backward(next(source.begin(), 7), next(source.begin(), 10), next(dest_backward.begin(), 5));
+        assert(dest_backward == result);
+        assert(res_copy_backward == next(dest_backward.begin(), 2));
+
+        vector<bool> dest_move(8, false);
+        const auto res_move = move(next(source.begin(), 7), next(source.begin(), 10), next(dest_move.begin(), 2));
+        assert(dest_move == result);
+        assert(res_move == next(dest_move.begin(), 5));
+
+        vector<bool> dest_move_backward(8, false);
+        const auto res_move_backward =
+            move_backward(next(source.begin(), 7), next(source.begin(), 10), next(dest_move_backward.begin(), 5));
+        assert(dest_move_backward == result);
+        assert(res_move_backward == next(dest_move_backward.begin(), 2));
+    }
+
+    { // sub char copy over char alignment dest
+        vector<bool> result = {false, false, false, false, false, false, false, true, false, true, false, false, false,
+            false, false, false};
+        vector<bool> dest(16, false);
+        const auto res_copy = copy(source.begin(), next(source.begin(), 3), next(dest.begin(), 7));
+        assert(dest == result);
+        assert(res_copy == next(dest.begin(), 10));
+
+        vector<bool> dest_n(16, false);
+        const auto res_copy_n = copy_n(source.begin(), 3, next(dest_n.begin(), 7));
+        assert(dest_n == result);
+        assert(res_copy_n == next(dest_n.begin(), 10));
+
+        vector<bool> dest_backward(16, false);
+        const auto res_copy_backward =
+            copy_backward(source.begin(), next(source.begin(), 3), next(dest_backward.begin(), 10));
+        assert(dest_backward == result);
+        assert(res_copy_backward == next(dest_backward.begin(), 7));
+
+        vector<bool> dest_move(16, false);
+        const auto res_move = move(source.begin(), next(source.begin(), 3), next(dest_move.begin(), 7));
+        assert(dest_move == result);
+        assert(res_move == next(dest_move.begin(), 10));
+
+        vector<bool> dest_move_backward(16, false);
+        const auto res_move_backward =
+            move_backward(source.begin(), next(source.begin(), 3), next(dest_move_backward.begin(), 10));
+        assert(dest_move_backward == result);
+        assert(res_move_backward == next(dest_move_backward.begin(), 7));
+    }
+}
+
+bool test_copy() {

```

Comment:
I'd like to see more exhaustive tests using `vector<unsigned char>` as a source of truth, it would probably be less code too. Additionally, we need to test overlapping ranges in both forward and reverse directions (only the ones permitted by the standard), since it would be easy for a mistake to result in writing extra stuff and then fixing it afterwards in a way that breaks overlapping ranges.

Reply:
I'm OK keeping the explicit tests in addition to the exhaustive ones though.

Reply:
we've caused regressions before (recently) because of performance improvements, so given we can exhaustively test things using we may as well. By exhaustively testing we mean the number of bits in the being copied and the offsets of the beginning and end iterators.

Reply:
I will likely add a randomized test since a truly exhaustive test could take a long time to execute.